File: Week_01/Practicum/Convert an Array Into a 2D Array With Conditions.cpp
// https://leetcode.com/problems/convert-an-array-into-a-2d-array-with-conditions/description/?fbclid=IwZXh0bgNhZW0CMTEAAR0gzMh_YHR0BpMHEfojkn83mKjjKDaBdkAAnz3B4CEb1_9It3Alsq8DBus_aem_-T_-RK2zI0rxH5r-7Do0sA

// O(nlogn) solution
/* Logic:
 *  Sort the array.
 *  For each number in the array, we need to check if the number is the same as the previous number.
 *  If it is, we add it to the next row.
 *  If it isn't, we reset the current number and add it to the first row.
 */
class Solution {
public:
    vector<vector<int>> findMatrix(vector<int>& nums) {
        vector<vector<int>> arr2D;

        sort(nums.begin(), nums.end());

        int currentNumber = -1;
        int currentNumberCount = 0;

        for (int i = 0; i < nums.size(); i++) {
            int num = nums[i];

            if (currentNumber == num){
                currentNumberCount ++;
            } else {
                currentNumber = num;
                currentNumberCount = 0;
            }

            // We need to initialize the vector for the current number if it doesn't exist.
            if(arr2D.size() < currentNumberCount + 1){
                arr2D.push_back(vector<int>());
            }

            arr2D[currentNumberCount].push_back(num);
        }

        return arr2D;
    }
};

File: Week_01/Practicum/python/Convert an Array Into a 2D Array With Conditions.py
class Solution:
    def findMatrix(self, nums: List[int]) -> List[List[int]]: 
        count_lst = [0 for _ in range(max(nums) + 1)]

        for num in nums:
            count_lst[num] += 1

        matrix = []

        for _ in range(max(count_lst)):
            row = []

            for i in range(len(count_lst)):
                if count_lst[i] != 0:
                    row.append(i)
                    count_lst[i] -= 1

            matrix.append(row)

        return matrix